### PR TITLE
Introduce `Param` class.

### PR DESCRIPTION
**Pull request by lucasb-eyer**
This fixes a lot of little annoyances I had over the past year or so. It also changes the interface of all layers with a bias as well as the `parameters` method. It cleans up the internal code considerably (e.g. `may_decay` disappears) and will allow to re-initialize an existing network, which is useful when memory is tight.

### CODE REVIEW

**Comment by lucasb-eyer** [comment]
I'll roll with it for a day or two and hit merge if nothing breaks until then.

File: DeepFried2/layers/BackwardsConvolutionCUDNN.py [hunk 1]
```diff
@@ -30,18 +29,20 @@ def __init__(self, nchan_in, nchan_out, filter_size, stride=1, border=0, mode='c
 
         w_shape = (nchan_in, nchan_out) + self.filter_size
         w_fan = (np.prod(self.filter_size)*nchan_out, np.prod(self.filter_size)*nchan_in)
+        w_name = ('Wconv_{},{}@{}' + 'x{}'*(len(w_shape) - 3)).format(*w_shape)
+        self.W = self._addparam(self.w_shape, init, fan=w_fan, name=w_name)
```
Comment (isaak-l) [inline DeepFried2/layers/BackwardsConvolutionCUDNN.py]: `self.w_shape` does not exist at this point. `w_shape` should be used here instead.

**Comment by lucasb-eyer** [comment]
Fixed in a force-push, see http://stackoverflow.com/questions/9813816/git-pull-after-forced-update for info on updating.

**Comment by lucasb-eyer** [comment]
Since both of us have been rolling with it for a few days without more problems, GO!

**Pull request merged**